### DISCUSSION TITLE
IdentityMap / Cascade Locking?

### DISCUSSION THREAD

**Discussion by parisholley** [mikro-orm/mikro-orm#4319] [General]
Somewhat related to #2380 as I don't think this would be a problem in most write requests/scenarios. I am noticing some strange behavior when entities with the same id are being populated with different values across two different `Promise`s. At times `Collection`s just simply disappear or a `Ref`s has been nulled out when accessed later. I suspect this may due to cascades happening across `Promise`s at the same time. It is hard to create a reproduction for as it isn't consistent due to latency differences/IO. It is more prevalent in serverless frameworks which execute nested routes in parallel.

As a workaround, when I am in a read-only request, I simply set `disableIdentityMap` to true (I wrap EntityManager with my own implementation) so that there are no collisions but this comes at a potential performance cost (I assume) as it can't shared pre-fetched entities across `Promise`s.

Digging through the code a bit, I think my suggestion would be to update the `EntityFactory.create` and introduce a queue here:

https://github.com/mikro-orm/mikro-orm/blob/ff001101fede72e3346a6700eb33a2fc3e4e510f/packages/core/src/entity/EntityFactory.ts#L55-L94

Every find is added to a queue with a callback `fn(existing:Entity) => Entity` which would prevent merges/cascades happening out of order.

**Comment by B4nan** [comment]
I don't think I understand, and I would like to see reproductions for the problem you have first, instead of suggesting some solution. At least a more verbose description with some pointers and code samples - you always say a few words and start discussing stuff as if I knew your codebase and the problems you deal with :] 

The propagation is synchronous, just like entity creation/hydration, so I am not sure how that could cause some inconsistencies? Also not sure what your actual proposal is, queue for what? The `ef.create` is a sync method, how can that be out of order, and what does an out-of-order merge/cascade even mean?

**Comment by parisholley** [comment]
@B4nan my hypothesis is not around async vs sync, node is single-threaded so that wouldn't cause a race condition. my concern is the event loop. are you saying with confidence that multiple calls to `findOne`/`find`/`populate` has no possibility of executing out of order? eg: there are no setTimeouts/IO/nextTick/event listeners? that subscribers/hooks or user logic couldn't cause the merge/cascade process to get pushed farther out on the event loop stack?

i'll keep digging to see if i can find areas of the framework that I can "break" and force an event loop mess, but it simply isn't practical to debug this and create a test case as the event loop can look drastically different on every request, hence trying to understand if architecturally there are places where we can add safe guards or if you have ideas on where sync execution can be delayed and I can focus on that area for reproducing.

**Reply by B4nan** [reply]
> there are no setTimeouts/IO/nextTick/event listeners

Nope, none that I am aware of, I hate ignoring promises or firing stuff without waiting for it in general.

> that subscribers/hooks or user logic couldn't cause the merge/cascade process to get pushed farther out on the event loop stack?

I still don't know what you mean by that - and I don't understand where you would want me to put the queue. But again, hooks are executed in order, serially, and awaited.

**Comment by parisholley** [comment]
still commenting out code to narrow down what I can, but at least a little more clarity, it seems to do with nested collections, here is the original code with the original error:

```typescript
    const fields = (
      await performable.fields.load({ populate: ['conditions.reference', 'values.conditions.field'] })
    ).toJSON();

    const dates = fields
        .map((f) => [
          f.lifecycle.created,
          f.lifecycle.archived,
          ...f.conditions.map((c) => c.lifecycle.created),
          ...f.conditions.map((c) => c.lifecycle.archived),
        ])
        .flat();
```

```
TypeError: Cannot read properties of undefined (reading 'map')
    at eval (/src/container/services/cache/FormCacheService.ts:51:23)
    at Array.map (<anonymous>)
```

my initial thought was perhaps doing toJSON() somehow held on to an old reference so i switched to library access like so:

```typescript
    const fields = (
      await performable.fields.load({ populate: ['conditions.reference', 'values.conditions.field'] })
    );

    const dates = fields.getItems()
        .map((f) => [
          f.lifecycle.created,
          f.lifecycle.archived,
          ...f.conditions.getItems().map((c) => c.lifecycle.created),
          ...f.conditions.getItems().map((c) => c.lifecycle.archived),
        ])
        .flat();
```

```
Error: Collection<PerformablePropertyCondition> of entity PerformablePropertyMikro[100] not initialized
    at Collection.checkInitialized (xxx/node_modules/.pnpm/@mikro-orm+core@6.0.0-dev.67/node_modules/@mikro-orm/core/entity/Collection.js:321:19)
    at Collection.getItems (xxx/node_modules/.pnpm/@mikro-orm+core@6.0.0-dev.67/node_modules/@mikro-orm/core/entity/Collection.js:99:18)
    at eval (/src/container/services/cache/FormCacheService.ts:51:23)
    at Array.map (<anonymous>)
```

going to keep digging, maybe we're lucky and its just a simple bug and not a race condition, but it is definately triggered from other EntityManager calls firing after I have awaited `performable.fields.load`

**Reply by parisholley** [reply]
another note here, is that the variance of the error being thrown goes down as the number of promises (entitymanager fires) happening at the same time is reduced (from me commenting them out), eg: every other request vs every request

**Comment by parisholley** [comment]
here is a reduced code reproduction (within my code base), i'll see if I can get it to throw an error within a test case:

```typescript

      await Promise.all([
        (async () => {
          const performable = await mikro.requireOne(
            PerformableMikro,
            { id: serviceId },
            { populate: ['vendor.rules'] }
          );

          const fields = await performable.fields.load({
            populate: ['conditions.reference', 'values.conditions.field'],
          });

          const dates = fields
            .getItems()
            .map((f) => [
              f.lifecycle.created,
              f.lifecycle.archived,
              ...f.conditions.getItems().map((c) => c.lifecycle.created),
              ...f.conditions.getItems().map((c) => c.lifecycle.archived),
            ])
            .flat();
        })(),
        (async () => {
          const performable = await mikro.requireOne(PerformableMikro, serviceId, {
            populate: ['fields.values', 'variants.values'],
          });
        })(),
      ]);
```

**Comment by parisholley** [comment]
@B4nan was able to reproduce :) PR coming shortly

**Reply by parisholley** [reply]
@B4nan see #4324

**Reply by B4nan** [reply]
I was digging a bit through older issues and found this one, and the PR that got autoclosed with the removal of v6 branch. Just wanted to mention that we have the opt-in dataloader implementation for loading collections and references now, and when you enable that in your test, it works fine consistently. The first test was already fixed even without that (I guess because we now use `em.populate` internally on more places), the third one needed `dataloader: true`.
